File: Python/Algorithm/Algorithm I/617. Merge Two Binary Trees.py
from typing import Optional


class Solution:
    def mergeTrees(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> Optional[TreeNode]:
        """
        Merges two binary trees by summing up corresponding nodes.

        Args:
            root1: The root of the first binary tree.
            root2: The root of the second binary tree.

        Returns:
            The root of the merged binary tree.

        Time Complexity: o(n), where n is the total number of nodes in the smaller of the two trees.
        Space Complexity: o(n), where n is the maximum height of the smaller of the two trees.
        """
        if root1 and root2:
            mergedTree = TreeNode(root1.val + root2.val)
        
            mergedTree.left = self.mergeTrees(root1.left, root2.left)
            mergedTree.right = self.mergeTrees(root1.right, root2.right)

        else:
            mergedTree = root1 if root1 else root2

        return mergedTree

File: Java/Algorithm/Algorithm I/617. Merge Two Binary Trees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

  /**
   * Merges two binary trees by summing the corresponding nodes.
   *
   * @param root1 The root of the first binary tree.
   * @param root2 The root of the second binary tree.
   * @return The merged binary tree.
   */
  public TreeNode mergeTrees(TreeNode root1, TreeNode root2) {
    if (root1 == null && root2 == null) return null;

    TreeNode mergedTree = new TreeNode();

    int root1Value = root1 != null ? root1.val : 0;
    int root2Value = root2 != null ? root2.val : 0;

    mergedTree.val = root1Value + root2Value;
    mergedTree.left =
      mergeTrees(
        root1 != null ? root1.left : null,
        root2 != null ? root2.left : null
      );
    mergedTree.right =
      mergeTrees(
        root1 != null ? root1.right : null,
        root2 != null ? root2.right : null
      );

    return mergedTree;
  }
}
